pmpro donations report

vbuster01 opened the pull request:
Add a widget on the PMPro Settings > Reports page to open a new donations report page when clicked. Reports can be filtered by year and/or month. Reports can be exported to CSV file.

### All Submissions:

* [x] Have you followed the [Contributing guidelines](https://github.com/strangerstudios/pmpro-pay-by-check/blob/dev/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/strangerstudios/pmpro-pay-by-check/pulls/) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Resolves XXX.

### How to test the changes in this Pull Request:

1.
2.
3.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

vbuster01 committed:
pmpro donations report

Add a widget on the PMPro Settings > Reports page to open a new donations report page when clicked. Reports can be filtered by year and/or month. Reports can be exported to CSV file.

clarnp042680 commented:
Would it be possible to add a option to select the user you want to view the Donation Report about?
I'm wanting to expand this to be able to choose a user to export a list of donations for a Year End Report for them. If I can get it to export to EXCEL for that user I can just pretty up the data in a template and issue it to them with letterhead. 
I have tried to manually change the code to select a ID but it didn't work. A quick dropdown with the members would be awesome to be able to select a quick export.

vbuster01 commented:
Hi.

I do not see why not. Let me have a look and see what I can do. Another questions. What type of financial year do you run?  Jan to Dec or Jul to Jun? The reason I asked is I have changed the initial code to work with a Jul to Jun  fiscal year.  I guess I could have a check box to select which type of year.

GRAHAM 
================================

On Nov 14, 2025, at 12:45, clarnp042680 ***@***.***> wrote:


clarnp042680
 left a comment 
(strangerstudios/pmpro-donations#83)
 <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3533878394>
Would it be possible to add a option to select the user you want to view the Donation Report about?
I'm wanting to expand this to be able to choose a user to export a list of donations for a Year End Report for them. If I can get it to export to EXCEL for that user I can just pretty up the data in a template and issue it to them with letterhead.
I have tried to manually change the code to select a ID but it didn't work. A quick dropdown with the members would be awesome to be able to select a quick export.

—
Reply to this email directly, view it on GitHub <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3533878394>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHGKABFVW7HAZ4PN4YNV2E334YIMXAVCNFSM6AAAAAB2O6CI72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZTHA3TQMZZGQ>.
You are receiving this because you authored the thread.

clarnp042680 commented:
Jan - Dec
A choice would just add even more coolness to an already good simple plugin... 

On November 14, 2025 1:26:06 PM EST, Graham Godfrey ***@***.***> wrote:
>vbuster01 left a comment (strangerstudios/pmpro-donations#83)
>
>Hi.
>
>I do not see why not. Let me have a look and see what I can do. Another questions. What type of financial year do you run?  Jan to Dec or Jul to Jun? The reason I asked is I have changed the initial code to work with a Jul to Jun  fiscal year.  I guess I could have a check box to select which type of year.
>
>GRAHAM 
>================================
>
>On Nov 14, 2025, at 12:45, clarnp042680 ***@***.***> wrote:
>
>
>clarnp042680
> left a comment 
>(strangerstudios/pmpro-donations#83)
> <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3533878394>
>Would it be possible to add a option to select the user you want to view the Donation Report about?
>I'm wanting to expand this to be able to choose a user to export a list of donations for a Year End Report for them. If I can get it to export to EXCEL for that user I can just pretty up the data in a template and issue it to them with letterhead.
>I have tried to manually change the code to select a ID but it didn't work. A quick dropdown with the members would be awesome to be able to select a quick export.
>
>—
>Reply to this email directly, view it on GitHub <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3533878394>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHGKABFVW7HAZ4PN4YNV2E334YIMXAVCNFSM6AAAAAB2O6CI72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZTHA3TQMZZGQ>.
>You are receiving this because you authored the thread.
>
>
>
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3534016608
>You are receiving this because you commented.
>
>Message ID: ***@***.***>

vbuster01 commented:
Here you go.
Updated to v3.
Features now include:
* Features:
 * - Display total donations received with detailed transaction history
 * - Filter by month and year (Fiscal Year: July-June or Calendar Year: Jan-Dec) - Default is Fiscal
 * - Filter by specific user email address
 * - View member details, including membership level
 * - Export filtered results to CSV
 * - Dashboard widget for quick access


https://github.com/vbuster01/delgensocorg/blob/main/pmpro-donations-report.php￼
delgensocorg/pmpro-donations-report.php at main · vbuster01/delgensocorg
github.com

Please let me know your feedback.

GRAHAM
================================

On Nov 14, 2025, at 13:31, clarnp042680 ***@***.***> wrote:


clarnp042680
 left a comment 
(strangerstudios/pmpro-donations#83)
 <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3534041871>Jan - Dec 
A choice would just add even more coolness to an already good simple plugin... 

On November 14, 2025 1:26:06 PM EST, Graham Godfrey ***@***.***> wrote: 
>vbuster01 left a comment (strangerstudios/pmpro-donations#83) 
> 
>Hi. 
> 
>I do not see why not. Let me have a look and see what I can do. Another questions. What type of financial year do you run? Jan to Dec or Jul to Jun? The reason I asked is I have changed the initial code to work with a Jul to Jun fiscal year. I guess I could have a check box to select which type of year. 
> 
>GRAHAM 
>================================ 
> 
>On Nov 14, 2025, at 12:45, clarnp042680 ***@***.***> wrote: 
> 
> 
>clarnp042680 
> left a comment 
>(strangerstudios/pmpro-donations#83) 
> <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3533878394> 
>Would it be possible to add a option to select the user you want to view the Donation Report about? 
>I'm wanting to expand this to be able to choose a user to export a list of donations for a Year End Report for them. If I can get it to export to EXCEL for that user I can just pretty up the data in a template and issue it to them with letterhead. 
>I have tried to manually change the code to select a ID but it didn't work. A quick dropdown with the members would be awesome to be able to select a quick export. 
> 
>— 
>Reply to this email directly, view it on GitHub <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3533878394>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHGKABFVW7HAZ4PN4YNV2E334YIMXAVCNFSM6AAAAAB2O6CI72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZTHA3TQMZZGQ>. 
>You are receiving this because you authored the thread. 
> 
> 
> 
> 
>-- 
>Reply to this email directly or view it on GitHub: 
>https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3534016608 
>You are receiving this because you commented. 
> 
>Message ID: ***@***.***>
—
Reply to this email directly, view it on GitHub <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3534041871>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHGKABA2QTWFOHTYOICJA7T34YNZJAVCNFSM6AAAAAB2O6CI72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZUGA2DCOBXGE>.
You are receiving this because you authored the thread.

vbuster01 commented:
I meant to add a screenshot.

￼

GRAHAM GODFREY
***@***.***
================================

On Nov 14, 2025, at 13:31, clarnp042680 ***@***.***> wrote:


clarnp042680
 left a comment 
(strangerstudios/pmpro-donations#83)
 <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3534041871>Jan - Dec 
A choice would just add even more coolness to an already good simple plugin... 

On November 14, 2025 1:26:06 PM EST, Graham Godfrey ***@***.***> wrote: 
>vbuster01 left a comment (strangerstudios/pmpro-donations#83) 
> 
>Hi. 
> 
>I do not see why not. Let me have a look and see what I can do. Another questions. What type of financial year do you run? Jan to Dec or Jul to Jun? The reason I asked is I have changed the initial code to work with a Jul to Jun fiscal year. I guess I could have a check box to select which type of year. 
> 
>GRAHAM 
>================================ 
> 
>On Nov 14, 2025, at 12:45, clarnp042680 ***@***.***> wrote: 
> 
> 
>clarnp042680 
> left a comment 
>(strangerstudios/pmpro-donations#83) 
> <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3533878394> 
>Would it be possible to add a option to select the user you want to view the Donation Report about? 
>I'm wanting to expand this to be able to choose a user to export a list of donations for a Year End Report for them. If I can get it to export to EXCEL for that user I can just pretty up the data in a template and issue it to them with letterhead. 
>I have tried to manually change the code to select a ID but it didn't work. A quick dropdown with the members would be awesome to be able to select a quick export. 
> 
>— 
>Reply to this email directly, view it on GitHub <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3533878394>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHGKABFVW7HAZ4PN4YNV2E334YIMXAVCNFSM6AAAAAB2O6CI72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZTHA3TQMZZGQ>. 
>You are receiving this because you authored the thread. 
> 
> 
> 
> 
>-- 
>Reply to this email directly or view it on GitHub: 
>https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3534016608 
>You are receiving this because you commented. 
> 
>Message ID: ***@***.***>
—
Reply to this email directly, view it on GitHub <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3534041871>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHGKABA2QTWFOHTYOICJA7T34YNZJAVCNFSM6AAAAAB2O6CI72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZUGA2DCOBXGE>.
You are receiving this because you authored the thread.

clarnp042680 commented:
Silly question since I'm just getting this after the holiday. I can't find where to update the file to enjoy the changes. I've checked everywhere.

vbuster01 commented:
No problems. My next step is to create a plugin that makes upgrades and installations easier.

Now I am making some assumptions here, because I don't know your site. 

I am assuming that you are using Code Snippets plugin from Code Snippets Pro. 

1. Find Snippets on the WordPress menu (Image attached) and click All Snippets. ![Snippets_Notification_Center](https://github.com/user-attachments/assets/afa6570a-e6b9-47c2-8b89-6fe10ce67666)
2. Deactivate the current snippet. Keep it as a backup if the new version blows up. You can delete it in a week or so, once you are happy that the new version works well.
3. At the top of the snippets page or in the menu, click Add New.
4. Give it a name like "PMPro-Donations-received-displayed-on-reports-page.-v3"
5. Assuming you have downloaded the new version, open it with whatever tool you use to view code, it could also be a text editor. 
6. Copy/paste all of the code into the new Code Snippet. 
7. Delete Line 1 of the code, the <?PHP statement as it is not required within the snippet. Line 1 will now be the start of the comment section, /**
8. Now, save and activate the code snippet.

Now check if the report displays without any problems. Remember, the default display is for a Fiscal display. Click the Calendar radio button to switch.

Let me know if you have any issues.
Graham

vbuster01 commented:
Hi

I replied on Github but it did not appear to come through as an email! Strange. I have copied/pasted my reply below.
============================================
No problems. My next step is to create a plugin that makes upgrades and installations easier.

Now I am making some assumptions here, because I don't know your site.

I am assuming that you are using Code Snippets plugin from Code Snippets Pro.

Find Snippets on the WordPress menu (Image attached) and click All Snippets.  <https://private-user-images.githubusercontent.com/30187524/521497424-afa6570a-e6b9-47c2-8b89-6fe10ce67666.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQ3NzUyNzcsIm5iZiI6MTc2NDc3NDk3NywicGF0aCI6Ii8zMDE4NzUyNC81MjE0OTc0MjQtYWZhNjU3MGEtZTZiOS00N2MyLThiODktNmZlMTBjZTY3NjY2LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjAzVDE1MTYxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAwOGY5ZGMxNmU3YjNjNGJhZWY1ZmJmODVkZDA2MDFkYTc2NmQxNGM3Y2QyOTVhMDNmZGM0MzA3YzMxMmY1M2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.KUHbw5r1oLwaFaX4r6965_FIi5YZkArEjQYhmeg8UCQ>
Deactivate the current snippet. Keep it as a backup if the new version blows up. You can delete it in a week or so, once you are happy that the new version works well.
At the top of the snippets page or in the menu, click Add New.
Give it a name like "PMPro-Donations-received-displayed-on-reports-page.-v3"
Assuming you have downloaded the new version, open it with whatever tool you use to view code, it could also be a text editor.
Copy/paste all of the code into the new Code Snippet.
Delete Line 1 of the code, the <?PHP statement as it is not required within the snippet. Line 1 will now be the start of the comment section, /**
Now, save and activate the code snippet.
Now check if the report displays without any problems. Remember, the default display is for a Fiscal display. Click the Calendar radio button to switch.

Let me know if you have any issues.

Graham

================================

On Dec 2, 2025, at 12:11, clarnp042680 ***@***.***> wrote:


clarnp042680
 left a comment 
(strangerstudios/pmpro-donations#83)
 <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3603124792>
Silly question since I'm just getting this after the holiday. I can't find where to update the file to enjoy the changes. I've checked everywhere.

—
Reply to this email directly, view it on GitHub <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3603124792>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHGKABA5LQ2SQCPAVJ453WD37XB43AVCNFSM6AAAAAB2O6CI72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMMBTGEZDINZZGI>.
You are receiving this because you authored the thread.

clarnp042680 commented:
Thanks for helping with that. My full time job has been missing the simple stuff sometimes. 

I like the drop down option but it's only showing my email address and user. I have over 100 others that I want to look at. Is it only showing me because of the Administrator tag for me? I am loving this option so far to get the report of the donations made just need it so I can select a particular user and export those records out quick.

Nate

vbuster01 commented:
Morning Nate

What user account are you logging into Wordpress with? It has to have Administrator or a Membership Manager privilege?  It must be administrator since you are creating code snippets.
Is this account the same account that may be associated with your membership but has the Administrator roll assigned?  I am just trying to figure out what is different?
On the all-users page, click on Administrators on the top. What is the Site Roll for your account? 

So to be clear. On the Donations report page, within the report you can see members that have donated, BUT in the pulldown to select a member there is only your email address?

Are the other filters working ok? Calendar year, Month, Year etc?

Can you supply a screen that will capture what is shown in the pulldown and with the members below. Please sanitize the screenshot.



Thanks

GRAHAM GODFREY

================================

On Dec 3, 2025, at 10:24, clarnp042680 ***@***.***> wrote:


clarnp042680
 left a comment 
(strangerstudios/pmpro-donations#83)
 <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3607392529>
Thanks for helping with that. My full time job has been missing the simple stuff sometimes.

I like the drop down option but it's only showing my email address and user. I have over 100 others that I want to look at. Is it only showing me because of the Administrator tag for me? I am loving this option so far to get the report of the donations made just need it so I can select a particular user and export those records out quick.

Nate

—
Reply to this email directly, view it on GitHub <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3607392529>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHGKABHCL6FMO7VDKIK4Z7L3736EDAVCNFSM6AAAAAB2O6CI72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMMBXGM4TENJSHE>.
You are receiving this because you authored the thread.

clarnp042680 commented:
I think I see my misunderstanding.
This report only shows the "extra donations" when a membership level is 
purchased not the actual membership donation.

On 12/3/2025 11:01 AM, Graham Godfrey wrote:
> *vbuster01* left a comment (strangerstudios/pmpro-donations#83) 
> <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3607575787>
> Morning Nate
>
> What user account are you logging into Wordpress with? It has to have 
> Administrator or a Membership Manager privilege? It must be 
> administrator since you are creating code snippets.
> Is this account the same account that may be associated with your 
> membership but has the Administrator roll assigned? I am just trying 
> to figure out what is different?
> On the all-users page, click on Administrators on the top. What is the 
> Site Roll for your account?
>
> So to be clear. On the Donations report page, within the report you 
> can see members that have donated, BUT in the pulldown to select a 
> member there is only your email address?
>
> Are the other filters working ok? Calendar year, Month, Year etc?
>
> Can you supply a screen that will capture what is shown in the 
> pulldown and with the members below. Please sanitize the screenshot.
>
>
>
> Thanks
>
> GRAHAM GODFREY
>
> ================================
>
> On Dec 3, 2025, at 10:24, clarnp042680 ***@***.***> wrote:
>
>
> clarnp042680
> left a comment
> (strangerstudios/pmpro-donations#83)
> <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3607392529> 
>
> Thanks for helping with that. My full time job has been missing the 
> simple stuff sometimes.
>
> I like the drop down option but it's only showing my email address and 
> user. I have over 100 others that I want to look at. Is it only 
> showing me because of the Administrator tag for me? I am loving this 
> option so far to get the report of the donations made just need it so 
> I can select a particular user and export those records out quick.
>
> Nate
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3607392529>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AHGKABHCL6FMO7VDKIK4Z7L3736EDAVCNFSM6AAAAAB2O6CI72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMMBXGM4TENJSHE>.
> You are receiving this because you authored the thread.
>
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3607575787>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AT2AHKUA5W7U3RZYWLKITQD374CMVAVCNFSM6AAAAAB2O6CI72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMMBXGU3TKNZYG4>.
> You are receiving this because you commented.Message ID: 
> ***@***.***>
>

vbuster01 commented:
Nate
Are ,yes. Let me look at it in the next day or so. You still have the original code snippet, correct? Just switch back to that one as it does not have that additional filtering.

What is the name of the website? Is there an acronym that I can use for the file name?

Thanks

GRAHAM GODFREY

================================

On Dec 3, 2025, at 11:52, clarnp042680 ***@***.***> wrote:


clarnp042680
 left a comment 
(strangerstudios/pmpro-donations#83)
 <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3607820500>I think I see my misunderstanding.
This report only shows the "extra donations" when a membership level is 
purchased not the actual membership donation.


<Snipped out old Message>

vbuster01 commented:
Nate

Are you using the PMPro Donations Add-on?
Do you have a Donations only membership level?  If so I have tested this a free account with donations enabled and it works fine. I see the test user and the correct amount displayed within the report. See image below.


	￼

What is your PMPro setup as there must be a difference! 
Do you have a paid memberships?

Can you send me a screenshot of your PMPro plugins if you want me to investigate further.

Thanks


GRAHAM GODFREY

================================

On Dec 3, 2025, at 22:53, GRAHAM GODFREY ***@***.***> wrote:

Nate
Are ,yes. Let me look at it in the next day or so. You still have the original code snippet, correct? Just switch back to that one as it does not have that additional filtering.

What is the name of the website? Is there an acronym that I can use for the file name?

Thanks

GRAHAM GODFREY

================================

On Dec 3, 2025, at 11:52, clarnp042680 ***@***.***> wrote:


clarnp042680
 left a comment 
(strangerstudios/pmpro-donations#83)
 <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3607820500>I think I see my misunderstanding.
This report only shows the "extra donations" when a membership level is 
purchased not the actual membership donation.


<Snipped out old Message>

vbuster01 commented:
Hopefully you have access to your database. Run the first simple SQL Query and let me know what output you have. You should see User ID, Email address, Donation amount and Date of the donation. There is a second query that will give you more detail.

-- 1. Query to show all donations with member IDs and email addresses
SELECT 
    o.user_id AS member_id,
    u.user_email,
    om.meta_value AS donation_amount,
    o.timestamp AS donation_date
FROM wp_pmpro_membership_ordermeta om
JOIN wp_pmpro_membership_orders o ON om.pmpro_membership_order_id = o.id
LEFT JOIN wp_users u ON o.user_id = u.ID
WHERE om.meta_key = 'donation_amount' 
    AND om.meta_value > 0
    AND u.user_email IS NOT NULL
ORDER BY o.timestamp DESC;

 
==================================================
-- 2. Query to show all donations with member IDs and email addresses with additional details
SELECT 
    o.user_id AS member_id,
    u.user_email,
    u.user_login,
    um1.meta_value AS first_name,
    um2.meta_value AS last_name,
    om.meta_value AS donation_amount,
    ml.name AS membership_level,
    o.timestamp AS donation_date
FROM wp_pmpro_membership_ordermeta om
JOIN wp_pmpro_membership_orders o ON om.pmpro_membership_order_id = o.id
LEFT JOIN wp_users u ON o.user_id = u.ID
LEFT JOIN wp_usermeta um1 ON u.ID = um1.user_id AND um1.meta_key = 'first_name'
LEFT JOIN wp_usermeta um2 ON u.ID = um2.user_id AND um2.meta_key = 'last_name'
LEFT JOIN wp_pmpro_membership_levels ml ON o.membership_id = ml.id
WHERE om.meta_key = 'donation_amount' 
    AND om.meta_value > 0
    AND u.user_email IS NOT NULL
ORDER BY o.timestamp DESC;

GRAHAM GODFREY

================================

On Dec 4, 2025, at 15:36, GRAHAM GODFREY ***@***.***> wrote:

Nate

Are you using the PMPro Donations Add-on?
Do you have a Donations only membership level?  If so I have tested this a free account with donations enabled and it works fine. I see the test user and the correct amount displayed within the report. See image below.


	<PastedGraphic-1.tiff>

What is your PMPro setup as there must be a difference! 
Do you have a paid memberships?

Can you send me a screenshot of your PMPro plugins if you want me to investigate further.

Thanks


GRAHAM GODFREY

================================

On Dec 3, 2025, at 22:53, GRAHAM GODFREY ***@***.***> wrote:

Nate
Are ,yes. Let me look at it in the next day or so. You still have the original code snippet, correct? Just switch back to that one as it does not have that additional filtering.

What is the name of the website? Is there an acronym that I can use for the file name?

Thanks

GRAHAM GODFREY

================================

On Dec 3, 2025, at 11:52, clarnp042680 ***@***.***> wrote:


clarnp042680
 left a comment 
(strangerstudios/pmpro-donations#83)
 <https://github.com/strangerstudios/pmpro-donations/pull/83#issuecomment-3607820500>I think I see my misunderstanding.
This report only shows the "extra donations" when a membership level is 
purchased not the actual membership donation.


<Snipped out old Message>

flintfromthebasement reviewed (changes requested):
**PR: strangerstudios/pmpro-donations#83 — pmpro donations report**
*vbuster01 (Graham Godfrey) → dev | 1 file, +306 -0*
<https://github.com/strangerstudios/pmpro-donations/pull/83>

**Summary**
This PR adds a new file `pmpro-donations-report.php` that provides a donations report page under PMPro Reports, with year/month filtering and CSV export. Not ready to merge — there's an XSS vector, a missing CSRF guard on the export, duplicate registration logic, and an architectural question about whether a snippet-style file belongs in this plugin at all.

---

**Issues**

**Critical** `pmpro-donations-report.php:233` — `add_query_arg()` output is echoed without `esc_url()`. This is a well-documented WordPress XSS vector: if the current URL contains injected input, `add_query_arg()` can pass it through unescaped.
```php
// current (unsafe)
<a href="<?php echo add_query_arg(array('export' => 'csv', 'month' => $month, 'year' => $year)); ?>">

// fix
<a href="<?php echo esc_url(add_query_arg(array('export' => 'csv', 'month' => $month, 'year' => $year))); ?>">
```

**Major** `pmpro-donations-report.php:103–130` (`MY_pmpro_donations_csv_export`) — The CSV export fires on `admin_init` based on `$_GET` params with no nonce verification. A crafted link can force an authenticated admin to trigger a data export. Add a nonce to the export link and verify it in the handler:
```php
// Generate link with nonce
esc_url(add_query_arg(array('export' => 'csv', ..., '_wpnonce' => wp_create_nonce('pmpro_donations_export'))))

// Verify in handler
if (!wp_verify_nonce($_GET['_wpnonce'] ?? '', 'pmpro_donations_export')) { wp_die(...); }
```

**Major** `pmpro-donations-report.php:97–99` — The report is registered twice: once via direct `global $pmpro_reports` mutation at file scope (line 97), and again via `add_filter('pmpro_reports', 'MY_pmpro_register_donations_report')` at line 303. The filter is the correct approach. The global mutation at file scope runs before hooks are ready and is redundant — remove it.

**Major** — This file is written as a code snippet (the header explicitly says "You can add this recipe to your site by creating a custom plugin or using the Code Snippets plugin"). It uses `MY_` as a placeholder prefix, has no `defined('ABSPATH') || exit;` guard, and isn't integrated with the existing `pmpro-donations.php` plugin code. If this is intended to ship as part of the pmpro-donations plugin, it needs to be integrated properly. If it's a recipe/snippet for the docs site, it doesn't belong in this repo at all.

**Minor** `pmpro-donations-report.php:36–57` — The base SQL query is built as a raw string and then `$wpdb->prepare()` is called on individually appended clauses. The callers do use `intval()` so the risk is mitigated, but the pattern is fragile — if these functions are ever called without `intval()` upstream, the prepare calls on appended clauses won't protect the base query. Collect all placeholders/values and call `$wpdb->prepare()` once on the full query.

**Minor** `pmpro-donations-report.php:183,225` — The `$` currency symbol is hardcoded throughout (display table and CSV output). PMPro has currency formatting helpers (`pmpro_formatPrice()` or `pmpro_getCurrencyText()`). Hard-coded `$` breaks installs using EUR, GBP, etc.

**Minor** `pmpro-donations-report.php:204` — Minimum year is hardcoded to 2022:
```php
for ($y = $current_year; $y >= 2022; $y--) {
```
This will silently drop older records from the filter dropdown as time passes. Derive the earliest year from the orders table instead.

---

**Questions**

1. Is this intended to be part of the pmpro-donations plugin itself, or a standalone snippet for the recipes/docs site? The file header and `MY_` prefix suggest the latter, but it's being added to the plugin repo.
2. The `pmpro_report_my_donations_widget()` and `pmpro_report_my_donations_page()` wrapper functions (lines 160, 291) exist solely to satisfy PMPro's naming convention but just call the `MY_` functions. Is that the expected integration pattern, or does the rename need to be consistent throughout?

andrewlimaza commented:
Closing this in favor of PR #94
